packages/core/src/declareAtom.ts

```diff
@@ -51,167 +76,123 @@ export function declareAtom<TState>(
     name = 'atom'
   }
 
-  const _id = nameToId(name as string | [TreeId])
+  const selfId = nameToId(name as string | [Id])
 
   if (initialState === undefined)
-    throwError(`Atom "${_id}". Initial state can't be undefined`)
+    throwError(`initial state (undefined) in atom "${selfId}"`)
 
-  const _tree = new Tree(_id)
-  const _deps = new Set<TreeId>()
-  // start from `0` for missing `actionDefault`
+  const selfNode = new Node(selfId)
+  const selfDeps = selfNode.deps
+  const selfDepsAll = selfNode.depsAll
+  const selfUpdates = selfNode.updates
   let dependencePosition = 0
   let initialPhase = true
 
-  function on<T>(
-    dep: Unit | PayloadActionCreator<T>,
-    reducer: Reducer<TState, T>,
-  ) {
+  function on<T>(dep: Unit<any>, reducer: Reducer<TState, T>) {
     if (!initialPhase)
-      throwError("Can't define dependencies after atom initialization")
-
-    safetyFunc(reducer, 'reducer')
+      throwError(`"on" handler calls after initialization in atom "${selfId}"`)
 
     const position = dependencePosition++
-    const depTree = getTree(dep as Unit)!
-    if (!depTree) throwError('Invalid dependency')
-    const depId = depTree.id
 
+    safetyFunc(reducer, 'atom reducer №' + position)
+
+    const depNode = getNode(dep)
+    if (!depNode) throwError('atom dependency №' + position)
+    const depId = depNode.id
     const isDepActionCreator = getIsAction(dep)
 
-    _tree.union(depTree)
+    depNode.depsAll.forEach(
+      id => !selfDepsAll.includes(id) && selfDepsAll.push(id),
+    )
```

Comment:
`selfDepsAll.push(...depNode.depsAll.filter(id => !selfDepsAll.includes(id)))`
must be better for perf, need test

Reply:
Why u think so? In my way we do not alocate new array